File: 3. Longest Substring Without Repeating Characters (Medium)/solution.go
package main

func longestSubstringLength(s string) int {
	lastOccurrence := make([]int, 256)
	for i := range lastOccurrence {
		lastOccurrence[i] = -1
	}
	start := 0
	maxLength := 0

	for i, char := range s {
		charCode := int(char)
		if lastOccurrence[charCode] >= start {
			start = lastOccurrence[charCode] + 1
		}
		lastOccurrence[charCode] = i
		if i-start+1 > maxLength {
			maxLength = i - start + 1
		}
	}

	return maxLength
}

func main() {
	println(longestSubstringLength("abcabcbb"))
	println(longestSubstringLength("bbbbb"))
	println(longestSubstringLength("pwwkew"))
}

File: 3. Longest Substring Without Repeating Characters (Medium)/solution.ts
function longestSubstringLength(s: string): number {
    const lastOccurrence: number[] = Array(256).fill(-1);
    let start = 0;
    let maxLength = 0;

    for (let i = 0; i < s.length; i++) {
        const charCode = s.charCodeAt(i);
        if (lastOccurrence[charCode] >= start) {
            start = lastOccurrence[charCode] + 1;
        }
        lastOccurrence[charCode] = i;
        maxLength = Math.max(maxLength, i - start + 1);
    }

    return maxLength;
}

console.log(longestSubstringLength("abcabcbb"));
console.log(longestSubstringLength("bbbbb"));
console.log(longestSubstringLength("pwwkew"));